ecl/hqlcpp/hqlhtcpp.cpp

```diff
@@ -10993,66 +11054,90 @@ ABoundActivity * HqlCppTranslator::doBuildActivityOutput(BuildCtx & ctx, IHqlExp
+            // assign the format based on the file type.
+            useGenericReadWrites = true;
+            IHqlExpression * fileType = queryAttributeChild(fileTypeOptionsExpr, fileTypeAtom, 0);
+            getStringValue(genericFileTypeFormat, fileType);
+            genericFileTypeFormat.toLowerCase();
+        }
+        else if (fileTypeOptionsExpr->getOperator() == no_pipe)
+        {
+            pipe = fileTypeOptionsExpr->queryChild(0);
+        }
     }
-    else if (filename->getOperator()==no_pipe)
+    else if (filename->getOperator() == no_pipe)
+    {
         pipe = filename->queryChild(0);
+    }
 
-    if (pipe && expr->hasAttribute(_disallowed_Atom))
-        throwError(HQLERR_PipeNotAllowed);
+    LinkedHqlExpr expireAttr = expr->queryAttribute(expireAtom);
+    IHqlExpression * seq = querySequence(expr);
 
     Owned<ABoundActivity> boundDataset = buildCachedActivity(ctx, dataset);
     ThorActivityKind kind = TAKdiskwrite;
     const char * activityArgName = "DiskWrite";
     SummaryType summaryType = SummaryType::WriteFile;
-    if (expr->getOperator() == no_spill)
-    {
-        kind = TAKspill;
-        activityArgName = "Spill";
-        summaryType = SummaryType::SpillFile;
-    }
-    else if (pipe)
-    {
-        kind = TAKpipewrite;
-        activityArgName = "PipeWrite";
-        summaryType = SummaryType::None;
-    }
-    else if (csvAttr)
-    {
-        kind = TAKcsvwrite;
-        activityArgName = "CsvWrite";
-    }
-    else if (xmlAttr)
-    {
-        kind = (isJson) ? TAKjsonwrite : TAKxmlwrite;
-        activityArgName = "XmlWrite";
-    }
-    else if (expr->hasAttribute(_spill_Atom))
+
+    IHqlExpression * csvAttr = expr->queryAttribute(csvAtom);
+    bool isJson = false;
+    IHqlExpression * xmlAttr = expr->queryAttribute(xmlAtom);
+
+    if (!useGenericReadWrites)
     {
-        kind = TAKspillwrite;
-        summaryType = SummaryType::SpillFile;
+        if (!xmlAttr)
+        {
+            xmlAttr = expr->queryAttribute(jsonAtom);
+            if (xmlAttr)
+                isJson = true;
+        }
+
+        if (pipe && expr->hasAttribute(_disallowed_Atom))
+            throwError(HQLERR_PipeNotAllowed);
+
+        if (expr->getOperator() == no_spill)
+        {
+            kind = TAKspill;
+            activityArgName = "Spill";
+            summaryType = SummaryType::SpillFile;
+        }
+        else if (pipe)
+        {
+            kind = TAKpipewrite;
+            activityArgName = "PipeWrite";
+            summaryType = SummaryType::None;
+        }
+        else if (csvAttr)
+        {
+            kind = TAKcsvwrite;
+            activityArgName = "CsvWrite";
+        }
+        else if (xmlAttr)
+        {
+            activityArgName = "XmlWrite";
+            if (isJson)
+                kind = TAKjsonwrite;
+            else
+                kind = TAKxmlwrite;
+        }
+        else if (expr->hasAttribute(_spill_Atom))
```

Comment:
I suspect this code and the following should be for both generic and non-generic.
Also this has been duplicated from lines 11106...

Reply:
The possible duplication was present before my changes. The second instance sets kind to TAKspillwrite, the first to TAKspill. Scanning the code, it looks like those two activity kinds are handled differently. Are these really duplicates?

I've removed my check for generic I/O, so all of these assignments will be checked either way.